synapse/storage/databases/main/event_federation.py

```diff
@@ -1583,6 +1583,43 @@ def _record_event_failed_pull_attempt_upsert_txn(
 
         txn.execute(sql, (room_id, event_id, 1, self._clock.time_msec(), cause))
 
+    @trace
+    async def get_event_ids_with_failed_pull_attempts(
+        self, event_ids: List[str]
+    ) -> List[str]:
+        """
+        Filter the given list of `event_ids` and return events which have any failed
+        pull attempts.
+
+        Args:
+            event_ids: A list of events to filter down.
+
+        Returns:
+            A filtered down list of `event_ids` that have previous failed pull attempts
+            (order is maintained).
+        """
+
+        rows = await self.db_pool.simple_select_many_batch(
+            table="event_failed_pull_attempts",
+            column="event_id",
+            iterable=event_ids,
+            keyvalues={},
+            retcols=("event_id",),
+            desc="get_event_ids_with_failed_pull_attempts",
+        )
+        event_ids_with_failed_pull_attempts_from_database = [
+            str(row["event_id"]) for row in rows
+        ]
+        # We want to maintain the order of the given `event_ids` so we re-construct the
+        # list since there is no gurantees from the database implementation/query.
+        event_ids_with_failed_pull_attempts = [
+            event_id
+            for event_id in event_ids
+            if event_id in event_ids_with_failed_pull_attempts_from_database
+        ]
```

Comment:
Why does the order matter, the caller just uses `x in event_ids_with_failed_pull_attempts`?

Reply:
In terms of why we want to try to process events in order `(depth, source order)`, this comment describes one reason why:

https://github.com/matrix-org/synapse/blob/c4e1533488e800786218b133590f3db78062f0bc/synapse/handlers/federation_event.py#L900-L901


It's also just more efficient to backfill this way (`depth` ascending) because one backfill event is likely to be the `prev_event` of the next event we're going to process. And source order absolutely matters for MSC2716 because many events are at the same depth so we use the tenuous sort that the other server gave us and hope they're doing their best.  The brittle nature of this is [one of the reasons why I don't want to continue on MSC2716](https://github.com/matrix-org/matrix-spec-proposals/pull/2716#issuecomment-1487441010) until we have online topological ordering.

TODO: Update comments with more of this context


---

In terms of whether `get_event_ids_with_failed_pull_attempts(...)` needs to do it. It seems like that detail doesn't matter in our current implementation as I have to go from `event_id` -> `event` in the downstream caller so end up re-constructing the list anyways.

But the implementation is probably interrelated to us trying to do something more efficient which is also discussed in this comment thread -> https://github.com/matrix-org/synapse/pull/15585#discussion_r1204479216

Reply:
I understand why we want to insert the events in order, but this isn't the order that we're inserting events; this is just a bundle of events that we're checking to see if we've failed before. (Aka it is a lookup table, not an ordered array.)

Reply:
Conclusion -> https://github.com/matrix-org/synapse/pull/15585#discussion_r1204683891

(basically, my bad and updated to `set()` without further work needed)